src/lock_manager.cc

```diff
@@ -51,3 +51,7 @@ void LockManager::Lock(const rocksdb::Slice &key) {
 void LockManager::UnLock(const rocksdb::Slice &key) {
   mutex_pool_[hash(key)]->unlock();
 }
+
+void LockManager::LockTwo(const rocksdb::Slice &first_key, const rocksdb::Slice &second_key) {
+  std::lock(*mutex_pool_[hash(first_key)], *mutex_pool_[hash(second_key)]);
```

Comment:
In my implementation, if the source and destination are the same LockManager::LockTwo won't be called. Instead, only one mutex will be locked via LockGuard. Additionally, `TwoLockGuard` or `LockTwo()` can be modified to check whether two keys are equal and make a proper lock. 
Yes, I was thinking about multi-lock, but there are 2 points:
- is it possible that more than 2 locks should be locked?
- multi-version should deal with templates :)

Reply:
Aha, we are using `crc(key)%buckets` to calculate the hash value, so different keys can't promise the value always be different.

>  Is it possible that more than 2 locks should be locked?

To be honest, I'm not sure. What I think is that `MultiLocks` contains 2 locks and more generic.

> multi-version should deal with templates :)

All is ok to me

Reply:
Unfortunately, I can't implement `MultiLockGuard` because of a lack of experience working with templates and variable argument functions. 

> Aha, we are using crc(key)%buckets to calculate the hash value, so different keys can't promise the value always be different.

In this case, I think, even in the case of implementing `MultiLockGuard` the issue of a possible deadlock will be present.

Reply:
My bad. They are two points:

1. Lock two keys without checking the hash value may cause deadlock
2. `MultiLockGuard` may be better than `TwoLockGuard` since it can be used in other scenarios

The current implementation also makes sense to me. Except we should check whether the hash value is same or not, only lock once if yes.

Reply:
I added a check that will track the case when two indexes from the mutex pool are the same. In that case, the second index will be increased to point to reference different mutex. Does it have any drawbacks?

Reply:
This way may unlock the wrong mutex? coz the unlock method didn't recalculate the mutex index. After reconsidering, to avoid deadlock that we need to do below things:

1. NOT acquire the same lock
2. Acquire locks with the hash index order, this may also cause deadlock if one thread acquired the `keys A B` and another acquired `B A`

Maybe I can try to implement multi locks.

Reply:
> This way may unlock the wrong mutex?

You are right. My bad.

Reply:
👍 It's a bit complex, I'm very happy to have a try and submit the PR to your branch.

Reply:
@torwig You can have a look at: https://github.com/torwig/kvrocks/pull/2

Reply:
@git-hulk Thanks! It looks like a good solution against a deadlock.